sound/soc/sof/sof-audio.c

```diff
@@ -116,8 +120,12 @@ int sof_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget)
 
 	trace_sof_widget_setup(swidget);
 
+	mutex_lock(&swidget->use_count_mutex);
+	swidget->use_count++;
+	mutex_unlock(&swidget->use_count_mutex);
+
 	/* widget already set up */
-	if (++swidget->use_count > 1)
+	if (swidget->use_count > 1)
```

Comment:
does this work?

Assuming use_count is zero, you could have a race where two threads increment use_count in a serial manner, and then both test the use_count as 2 and both return without doing anything. I think you need to protect the rest, not just the incrementation of the variable.

Reply:
yes, I think the both sof_widget-setup() and free() must be protected as a whole too

Reply:
The threads which can increase/decrease are protected by higher level locking in PCM level, we need to protect this against the checks happening outside of the PCM lock, like the kcontrol access.

Reply:
Right, hrm, there might be concurrent PCM streams touching the same widget? That might be an issue, but that issue already exists, no? The chance is slim, but in theory it could happen, the `use_count` is not atomic...

Reply:
Blindly protecting these function as whole results (when running aplay):
```
[13115.199981] ================================================
[13115.199984] WARNING: lock held when returning to user space!
[13115.199988] 6.1.0-rc6+ #841 Not tainted
[13115.199993] ------------------------------------------------
[13115.199995] aplay/13597 is leaving the kernel with locks still held!
[13115.199997] 2 locks held by aplay/13597:
[13115.200000]  #0: ffff9a1304375c88 (&swidget->use_count_mutex){....}-{3:3}, at: sof_widget_setup+0x84/0x3a0 [snd_sof]
[13115.200038]  #1: ffff9a1304374e88 (&swidget->use_count_mutex){....}-{3:3}, at: sof_widget_setup+0x84/0x3a0 [snd_sof]
```
Needs to be narrowed down or to protect the increment/decrement + the test.

Reply:
I suppose protecting the use_count during control IO is good enough then

Reply:
Can we not have multiple controls using the same widget? I think this needs a lot more explanations on what agents will simultaneously use swidget->use_count

Reply:
yes, we can but the mutex would protect all the get/put ops for all the kcontrols right?

Reply:
the other question I would have is that if we only want to protect the kcontrol path, and the incrementation and decrementation of a variable is really protected by a PCM mutex, isn't a spin_lock good enough and the mutex overkill?

Reply:
why do you think that @plbossart. I'd think its the other way around.

Reply:
besides a spinlock cannot be used for the purpose of this PR at all because we'd have to sleep while waiting for a reply for set_get_data IPC and a spinlock is meant to be atomic right?

Reply:
you're right @ranj063 not sure what I was thinking